src/scp/SCP.cpp

```diff
@@ -120,28 +120,93 @@ SCP::getJsonInfo(size_t limit, bool fullKeys)
     return ret;
 }
 
+SCP::QuorumInfoNodeState
+SCP::getState(NodeID const& node, uint64 slotIndex, int N = 2)
+{
+    for (int k = 0; k < N; k++)
+    {
+        if (slotIndex - k <= 0)
+        {
+            break;
+        }
+        auto slot = getSlot(slotIndex - k, false);
+        if (slot)
+        {
+            bool selfAlreadyMovedOn = k > 0;
+            auto s = slot->getState(node, selfAlreadyMovedOn);
+            if (s != SCP::QuorumInfoNodeState::NO_INFO)
+            {
+                return s;
+            }
+        }
+    }
+    // We have no messages from this node in N slots.
+    // At this point, we'll consider this node MISSING.
+    return SCP::QuorumInfoNodeState::MISSING;
+}
+
 Json::Value
 SCP::getJsonQuorumInfo(NodeID const& id, bool summary, bool fullKeys,
                        uint64 index)
 {
     Json::Value ret;
+    if (mKnownSlots.empty())
+    {
+        return ret;
+    }
     if (index == 0)
     {
-        for (auto& item : mKnownSlots)
-        {
-            auto& slot = *item.second;
-            ret = slot.getJsonQuorumInfo(id, summary, fullKeys);
-            ret["ledger"] = static_cast<Json::UInt64>(slot.getSlotIndex());
-        }
+        index = mKnownSlots.rbegin()->first;
```

Comment:
Interesting the old code was looping over all slots only to keep the last result... I am pretty sure that was not what we intended to do originally (was returning all slots) - and indeed I found that https://github.com/stellar/stellar-core/pull/2127 made that change.
I think this is fine. Funny that it took 3 years to notice.